### PR TITLE
machine-types/core refactoring to make it more flexible

### PR DESCRIPTION
**Pull request by jouvin**
- Initial part factored out in a separate template, core-init.pan, to allow
execution of site specific actions between this initial part and the core configuration
- Ability to delay some part of the configuration, like file systems, in case some site
information must be defined before
- Hook to configure time synchronisation according to site needs (ntp, chrony...)
- Template cleanup

### CODE REVIEW

File: os/kernel_version_arch.pan [hunk 1]
```diff
@@ -141,51 +151,58 @@ variable KERNEL_EXPLICITLY_DEFINED = is_defined(KERNEL_VERSION_NUM) || is_define
...
-    } else if ( is_defined(OS_VERSION_PARAMS['version']) &&
-                is_defined(OS_KERNEL_VERSION[OS_VERSION_PARAMS['version']]) ) {
-      OS_KERNEL_VERSION[OS_VERSION_PARAMS['version']];
+    if ( is_defined(NODE_OS_VERSION) ) {
+        if ( is_defined(OS_KERNEL_VERSION[NODE_OS_VERSION]) ) {
+            # Kernel version specific to version+arch
+            OS_KERNEL_VERSION[NODE_OS_VERSION];
+        } else if ( is_defined(KERNEL_OS_VERSION_PARAMS['version']) &&
+                    is_defined(OS_KERNEL_VERSION[KERNEL_OS_VERSION_PARAMS['version']]) ) {
+            OS_KERNEL_VERSION[KERNEL_OS_VERSION_PARAMS['version']];
```
Comment (Pansanel) [inline os/kernel_version_arch.pan:160]:
```
        } else if ( is_defined(KERNEL_OS_VERSION_PARAMS['version'])
                && is_defined(OS_KERNEL_VERSION[KERNEL_OS_VERSION_PARAMS['version']]) ) {
            OS_KERNEL_VERSION[KERNEL_OS_VERSION_PARAMS['version']];
```

File: os/kernel_version_arch.pan [hunk 2]
```diff
@@ -6,126 +6,136 @@ variable KERNEL_VERSION_CONFIG ?= null;
...
-  'sl580', '2.6.18-308.1.1.el5',
-  'sl600', '2.6.32-71.el6',
-  'sl610', '2.6.32-131.0.15.el6',
-  'sl620', '2.6.32-220.el6',
-  'sl630', '2.6.32-279.el6',
-  'sl640', '2.6.32-358.el6',
-  'centos550', '2.6.18-194.el5',
-  'fedora14', '2.6.35.6-45.fc14',
-  'opensuse11','2.6.34-12.3',
+variable OS_KERNEL_VERSION_DEFAULT ?= dict(
```
[the remark this answers is not in the document]
Comment (jouvin) [reply os/kernel_version_arch.pan:75]: You are right that some cleanup is required in this template to remove all the errata managenement and this kind of things predating the use of YUM... But to keep the things as simple as possible, I'd prefer to do it in a separate PR. Feel free to open an issue for that and assign it to me! For this PR, I just remove the obsolete entries in the dict.

File: os/kernel_version_arch.pan [hunk 3]
```diff
@@ -141,51 +151,58 @@ variable KERNEL_EXPLICITLY_DEFINED = is_defined(KERNEL_VERSION_NUM) || is_define
...
-      if ( is_defined(v["size"]) ) {
-        ram_size = ram_size + v["size"];
-      } else {
-        error('RAM size undefined for bank '+to_string(0));
-      };
-    };
-    if ( (KERNEL_SMP_PARAMS['limit'] > 0) && (ram_size >= KERNEL_SMP_PARAMS['limit'])) {
-      debug('Memory size above the smp kernel limit. Using largesmp kernel.');
-      KERNEL_SMP_PARAMS['largesmp'];
+    if ( is_defined(KERNEL_OS_VERSION_PARAMS['version']) && (KERNEL_OS_VERSION_PARAMS['major'] == 'sl4') ) {
```
[the remark this answers is not in the document]
Comment (jouvin) [reply os/kernel_version_arch.pan:183]: Removed. As previously, probably need a deeper cleanup to remove the `KERNEL_VARIANT` variable which is probably useless now...

File: os/kernel_version_arch.pan [hunk 4]
```diff
@@ -6,126 +6,136 @@ variable KERNEL_VERSION_CONFIG ?= null;
...
+                # in OS specific template config/os/kernel_version_arch.tpl.
+                arch = 'i686';
+            };
+        } else {
+            error('Distribution architecture not found, unable to guess OS architecture used');
+        };
+    };
+
+    if ( exists("/hardware/cpu/0/arch") &&
+        (value("/hardware/cpu/0/arch") == 'i386') &&
```
[the remark this answers is not in the document]
Comment (jouvin) [reply os/kernel_version_arch.pan:57]: Agreed

File: os/kernel_version_arch.pan (hunk 1 above)
Comment (jouvin) [reply os/kernel_version_arch.pan:160]: Agreed

File: os/version.pan [hunk 5]
```diff
@@ -13,98 +13,100 @@ variable NODE_OS_VERSION_DEFAULT ?= undef;
...
-                                 };
-
 # Define GLITE_CONFIG_HACKS if required by selected OS version
 variable GLITE_CONFIG_HACKS = if ( is_defined(NODE_OS_VERSION) &&
-                                   is_defined(GLITE_CONFIG_HACKS_DB[escape(NODE_OS_VERSION)]) ) {
-                                GLITE_CONFIG_HACKS_DB[escape(NODE_OS_VERSION)];
-                              } else {
-                                SELF;
-                              };
+                                    is_defined(GLITE_CONFIG_HACKS_DB[escape(NODE_OS_VERSION)]) ) {
```
[the remark this answers is not in the document]
Comment (jouvin) [reply os/version.pan:103]: Agreed

File: os/version.pan [hunk 6]
```diff
@@ -13,98 +13,100 @@ variable NODE_OS_VERSION_DEFAULT ?= undef;
...
-  };
+    };
 };
 
 variable OS_FLAVOUR ?= {
-    if ( is_defined(OS_VERSION_PARAMS['flavour']) ) {
-      OS_VERSION_PARAMS['major'] +"."+OS_VERSION_PARAMS['flavour']+"-"+ OS_VERSION_PARAMS['arch'];
+    if ( is_defined(OS_VERSION_PARAMS_DEFAULT['flavour']) ) {
+        OS_VERSION_PARAMS_DEFAULT['major'] + "." + OS_VERSION_PARAMS_DEFAULT['flavour'] +
+                                                        "-" + OS_VERSION_PARAMS_DEFAULT['arch'];
```
[the remark this answers is not in the document]
Comment (jouvin) [reply os/version.pan:80]: Agreed

**Comment by jouvin** [comment]
@Pansanel thanks for your detailed review and apologizes for the late follow-up... Hopefully I fixed everything.

File: os/kernel_version_arch.pan (hunk 2 above)
[the remark this answers is not in the document]
Comment (jrha) [reply os/kernel_version_arch.pan:75]: I've opened issue #152 for this.

**Review by jrha — APPROVED** [review]
LGTM

**Pull request merged**